### PR TITLE
Use Val{} to separate between types of problems 

### PR DESCRIPTION
**Pull request by abelsiqueira**
Rebase after #22

### CODE REVIEW

**Comment by abelsiqueira** [comment]
cf. @dpo - Can't ask for the review through the button.

File: src/al.jl [hunk 1]
```diff
@@ -4,40 +4,52 @@ using Logging, SolverTools, NLPModels
...
 
-function al(nlp :: AbstractNLPModel; mu :: Real = eltype(nlp.meta.x0)(10.0),
+function al(::Val{:tron}, nlp :: AbstractNLPModel;
             max_iter :: Int = 1000, max_time :: Real = 30.0, max_eval :: Int=-1,
-            atol :: Real = 1e-7, rtol :: Real = 1e-7,
+            atol :: Real = 1e-8, rtol :: Real = 1e-8,
             subsolver_logger :: AbstractLogger=NullLogger(),
            )
+  if !(unconstrained(nlp) || bound_constrained(nlp))
+    error("al(::Val{:tron}, nlp) should be called for unconstrained or bound-constrained problems. Use al(nlp)")
```
Comment (dpo) [inline src/al.jl:23]: "should only be called"

File: src/al.jl [hunk 2]
```diff
@@ -4,40 +4,52 @@ using Logging, SolverTools, NLPModels
...
+  @warn "Problem doesn't have constraints, so tron is being called instead"
+  return tron(nlp, subsolver_logger=subsolver_logger, atol=atol, rtol=rtol, max_eval=max_eval, max_time=max_time)
+end
 
-  T = eltype(nlp.meta.x0)
-
-  if nlp.meta.ncon == 0 # unconstrained
+function al(::Val{:ineq}, nlp :: AbstractNLPModel; kwargs...)
+  if nlp.meta.ncon == 0 || equality_constrained(nlp)
+    error("al(::Val{:ineq}, nlp) should be called for problems with inequalities. Use al(nlp)")
```
Comment (dpo) [inline src/al.jl:31]: Same.

File: src/al.jl [hunk 3]
```diff
@@ -4,40 +4,52 @@ using Logging, SolverTools, NLPModels
...
+  snlp = SlackModel(nlp)
+  output = al(Val(:equ), snlp; kwargs...)
+  output.solution = output.solution[1:nlp.meta.nvar]
+  return output
+end
 
-    S = with_logger(subsolver_logger) do
-      tron(nlp)
-    end
+"""Implementation of a augmented Lagrangian method for:
```
Comment (dpo) [inline src/al.jl:40]: an

File: src/al.jl [hunk 4]
```diff
@@ -4,40 +4,52 @@ using Logging, SolverTools, NLPModels
...
-    return GenericExecutionStats(status, nlp, solution = x, objective = obj(nlp, x), dual_feas = normgp, primal_feas = normcx,
-                                 iter = iter, elapsed_time = el_time)
+function al(::Val{:equ}, nlp :: AbstractNLPModel; mu :: Real = eltype(nlp.meta.x0)(10.0),
+            max_iter :: Int = 1000, max_time :: Real = 30.0, max_eval :: Int=-1,
+            atol :: Real = 1e-8, rtol :: Real = 1e-8,
+            subsolver_logger :: AbstractLogger=NullLogger(),
+           )
+  if nlp.meta.ncon == 0 || !equality_constrained(nlp)
+    error("al(::Val{:equ}, nlp) should be called for equality-constrained problems. Use al(nlp)")
   end
```
Comment (dpo) [inline src/al.jl:50]: Here the problem could also have bounds right?

File: src/al.jl [hunk 5]
```diff
@@ -120,6 +132,7 @@ function al(nlp :: AbstractNLPModel; mu :: Real = eltype(nlp.meta.x0)(10.0),
     end
   end
 
-  return GenericExecutionStats(status, nlp, solution = al_nlp.x[1:nlp.meta.nvar-ns], objective = obj(nlp, al_nlp.x), dual_feas = normgp, primal_feas = normcx,
-                               iter = iter, elapsed_time = el_time)
+  return GenericExecutionStats(status, nlp, solution = al_nlp.x,
+                               objective = obj(nlp, al_nlp.x), dual_feas = normgp, primal_feas = normcx,
```
Comment (dpo) [inline src/al.jl:136]: Maybe the augmented Lagrangian struct could keep the value of the original obj around so it doesn't have to be recomputed?

File: src/al.jl (hunk 4 above)
Comment (abelsiqueira) [reply src/al.jl:50]: Right, I'm gonna add "[problems] with bounded variables."

File: src/al.jl (hunk 5 above)
Comment (abelsiqueira) [reply src/al.jl:136]: I think `S.objective` can reused removing the aug. penalty part.

Comment (abelsiqueira) [reply src/al.jl:136]: But I think it's better in a new PR.

File: src/al.jl [hunk 6]
```diff
@@ -4,40 +4,52 @@ using Logging, SolverTools, NLPModels
...
+function al(::Val{:tron}, nlp :: AbstractNLPModel;
             max_iter :: Int = 1000, max_time :: Real = 30.0, max_eval :: Int=-1,
-            atol :: Real = 1e-7, rtol :: Real = 1e-7,
+            atol :: Real = 1e-8, rtol :: Real = 1e-8,
             subsolver_logger :: AbstractLogger=NullLogger(),
            )
+  if !(unconstrained(nlp) || bound_constrained(nlp))
+    error("al(::Val{:tron}, nlp) should only be called for unconstrained or bound-constrained problems. Use al(nlp)")
+  end
+  @warn "Problem doesn't have constraints, so tron is being called instead"
```
Comment (dpo) [inline src/al.jl:25]: instead of what? Here I would write "problem does not have general constraints; calling tron.".

File: src/al.jl (hunk 5 above)
Comment (dpo) [reply src/al.jl:136]: That's right. You're already doing something similar on line 35, so might as well do it here, no?

Comment (abelsiqueira) [reply src/al.jl]: Ok. It's updated.

**Comment by abelsiqueira** [comment]
This shall be fixed soon when I move it to JSO.

**Pull request merged**